File: py/non-inclusive/cni.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func loadInclusiveMapping(file *os.File, delim string) map[string]string {
	inclusiveMapping := make(map[string]string)

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, delim)
		if len(parts) == 2 {
			nonInclusive := strings.TrimSpace(strings.ToLower(parts[0]))
			inclusive := strings.TrimSpace(parts[1])
			inclusiveMapping[nonInclusive] = inclusive
		}
	}

	return inclusiveMapping
}

func replaceNonInclusive(document string, inclusiveMapping map[string]string) string {
	words := strings.Fields(document)
	for i, word := range words {
		lowerWord := strings.ToLower(word)
		if inclusive, found := inclusiveMapping[lowerWord]; found {
			words[i] = "[" + word + "/" + inclusive + "]"
		}
	}
	return strings.Join(words, " ")
}

func main() {
	if len(os.Args) < 4 {
		fmt.Println("Usage: go run main.go --noninclusive <non_inclusive_file> --doc <document_file> [--delim <delimiter>]")
		os.Exit(1)
	}

	nonInclusiveFile, err := os.Open(os.Args[2])
	if err != nil {
		fmt.Println("Error: Unable to open non-inclusive words file.", err)
		os.Exit(1)
	}
	defer nonInclusiveFile.Close()

	documentFile, err := os.Open(os.Args[3])
	if err != nil {
		fmt.Println("Error: Unable to open document file.", err)
		os.Exit(1)
	}
	defer documentFile.Close()

	delim := ","
	if len(os.Args) >= 6 && os.Args[4] == "--delim" {
		delim = os.Args[5]
	}

	inclusiveMapping := loadInclusiveMapping(nonInclusiveFile, delim)

	scanner := bufio.NewScanner(documentFile)
	var document string
	for scanner.Scan() {
		document += scanner.Text() + "\n"
	}

	replacedDocument := replaceNonInclusive(document, inclusiveMapping)

	fmt.Println("Replaced document:")
	fmt.Println(replacedDocument)
}

File: py/non-inclusive/cni.rs
use std::collections::HashMap;
use std::env;
use std::error::Error;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::path::Path;

fn load_inclusive_mapping(file: &File, delim: &str) -> HashMap<String, String> {
    let mut inclusive_mapping = HashMap::new();
    let reader = BufReader::new(file);

    for line in reader.lines() {
        if let Ok(line) = line {
            let parts: Vec<&str> = line.trim().split(delim).collect();
            if parts.len() == 2 {
                let non_inclusive = parts[0].trim().to_lowercase();
                let inclusive = parts[1].trim().to_string();
                inclusive_mapping.insert(non_inclusive, inclusive);
            }
        }
    }

    inclusive_mapping
}

fn replace_non_inclusive(document: &str, inclusive_mapping: &HashMap<String, String>) -> String {
    document
        .split_whitespace()
        .map(|word| {
            let lower_word = word.to_lowercase();
            if let Some(inclusive) = inclusive_mapping.get(&lower_word) {
                format!("[{}]", inclusive)
            } else {
                word.to_string()
            }
        })
        .collect::<Vec<String>>()
        .join(" ")
}

fn main() -> Result<(), Box<dyn Error>> {
    let args: Vec<String> = env::args().collect();

    if args.len() < 4 {
        eprintln!("Usage: cargo run -- --noninclusive <non_inclusive_file> --doc <document_file> [--delim <delimiter>]");
        return Ok(());
    }

    let non_inclusive_path = &args[2];
    let document_path = &args[3];
    let delim = if args.len() >= 6 && args[4] == "--delim" {
        args[5].as_str()
    } else {
        ","
    };

    let non_inclusive_file = File::open(non_inclusive_path)?;
    let inclusive_mapping = load_inclusive_mapping(&non_inclusive_file, delim);

    let document_file = File::open(document_path)?;
    let document_reader = BufReader::new(document_file);
    let document: Vec<String> = document_reader.lines().collect::<Result<_, _>>()?;
    let document_str = document.join("\n");

    let replaced_document = replace_non_inclusive(&document_str, &inclusive_mapping);

    println!("Replaced document:");
    println!("{}", replaced_document);

    Ok(())
}

